src/singleton_snap.py

```diff
@@ -0,0 +1,293 @@
+        try:
+            os.makedirs(self.LOCK_DIR, exist_ok=True)
+            os.chown(self.LOCK_DIR, os.geteuid(), os.getegid())
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise SingletonSnapError(f"Error creating lock directory: {e}") from e
+
+    def _get_registration_file_path(self, snap_name: str) -> str:
+        """Helper method to construct registration file path.
+
+        The registration file name is constructed as:
+        LCK..{snap_name}--{unit_name}
+        """
+        filename = f"{self.LOCK_FILE_PREFIX}{snap_name}{self.SEPARATOR}{self.unit_name}"
+        return os.path.join(self.LOCK_DIR, filename)
+
+    def _parse_unit_name_from_file(self, filename: str, snap_name: str) -> str:
+        """Helper method to extract unit name from registration filename."""
+        prefix = f"{self.LOCK_FILE_PREFIX}{snap_name}{self.SEPARATOR}"
+        return filename[len(prefix) :]
+
+    def register(self, snap_name: str) -> None:
+        """Register current unit as using the specified snap.
+
+        Raises:
+            SingletonSnapError: if there is an I/O related error creating the lock file.
+        """
+        self._update_registration(snap_name, create=True)
+
+    def unregister(self, snap_name: str) -> None:
+        """Unregister current unit from using the specified snap.
+
+        Raises:
+            SingletonSnapError: if there is an I/O related error removing the lock file.
+        """
+        self._update_registration(snap_name, create=False)
+
+    @contextmanager
+    def _lock_directory(self):
+        dir_fd = None
+        try:
+            dir_fd = os.open(self.LOCK_DIR, os.O_RDONLY)
+            fcntl.flock(dir_fd, fcntl.LOCK_SH)
+            yield
+        except TypeError as e:
+            raise SingletonSnapError(f"Invalid lock directory path: {e}") from e
+        except FileNotFoundError as e:
+            raise SingletonSnapError(f"Lock directory not found: {e}") from e
+        except OSError as e:
+            raise SingletonSnapError(f"Error locking directory: {e}") from e
+        finally:
+            if dir_fd is not None:
+                try:
+                    fcntl.flock(dir_fd, fcntl.LOCK_UN)
+                    os.close(dir_fd)
+                except Exception as e:
+                    raise SingletonSnapError(f"Failed to release lock: {e}") from e
+
+    def _update_registration(self, snap_name: str, create: bool) -> None:
+        """Internal method to handle registration or unregistration of a snap.
+
+        Achieved by creating or removing a lock file; atomic operation is ensured
+        with a directory lock.
+
+        Raises:
+            SingletonSnapError: If an OS error occurs.
+        """
+        lock_path = self._get_registration_file_path(snap_name)
+        action = "registering" if create else "unregistering"
+
+        try:
+            with self._lock_directory():
+                if create:
```

Comment:
This function has two different modes and if/else between them (another if/else in the `action = ...` line above). Would it be better to just include the correct code directly in register/unregister?

Reply:
I had included that in the first draft, then refactored it into this version to simplify it, but I agree that it now has too many levels of indentation, which might not be an improvement in terms of readability. I will change it back.